The job **`dist-x86_64-linux-alt`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/4548216454/jobs/8019021733) [(plain)](https://github.com/rust-lang-ci/rust/commit/dcc6fdf55c51964a5fef0f0e5cfcb28d013a1bf2/checks/12348844256/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Building stage1 library artifacts (x86_64-unknown-linux-gnu) 
error: failed to run `rustc` to learn about target-specific information

Caused by:
  process didn't exit successfully: `/checkout/obj/build/bootstrap/debug/rustc - --crate-name ___ --print=file-names -Csymbol-mangling-version=legacy -Zunstable-options -Zunstable-options '--check-cfg=values(bootstrap)' '--check-cfg=values(stdarch_intel_sde)' '--check-cfg=values(no_fp_fmt_parse)' '--check-cfg=values(no_global_oom_handling)' '--check-cfg=values(no_rc)' '--check-cfg=values(no_sync)' '--check-cfg=values(freebsd12)' '--check-cfg=values(freebsd13)' '--check-cfg=values(backtrace_in_libstd)' '--check-cfg=values(target_env,"libnx")' '--check-cfg=values(target_arch,"asmjs","spirv","nvptx","xtensa")' -Zmacro-backtrace -Clink-args=-Wl,-z,origin '-Clink-args=-Wl,-rpath,$ORIGIN/../lib' -Clink-args=-fuse-ld=lld -Csplit-debuginfo=off -Cprefer-dynamic -Zinline-mir -Cembed-bitcode=yes '-Zcrate-attr=doc(html_root_url="https://doc.rust-lang.org/nightly/")' --target x86_64-unknown-linux-gnu --crate-type bin --crate-type rlib --crate-type dylib --crate-type cdylib --crate-type staticlib --crate-type proc-macro --print=sysroot --print=split-debuginfo --print=crate-name --print=cfg` (exit status: 254)
  --- stderr
  /checkout/obj/build/x86_64-unknown-linux-gnu/stage1/lib/librustc_driver-d8b5aaf433c6b018.so(+0x125d583)[0x7fb817915583]
  /lib64/libpthread.so.0(+0xf630)[0x7fb81609f630]
  /checkout/obj/build/x86_64-unknown-linux-gnu/stage1/bin/rustc(+0x1c33f)[0x55b01deb133f]
  /checkout/obj/build/x86_64-unknown-linux-gnu/stage1/lib/librustc_driver-d8b5aaf433c6b018.so(+0x12a02ee)[0x7fb8179582ee]
  /checkout/obj/build/x86_64-unknown-linux-gnu/stage1/lib/librustc_driver-d8b5aaf433c6b018.so(+0x129a63f)[0x7fb81795263f]
  /checkout/obj/build/x86_64-unknown-linux-gnu/stage1/lib/librustc_driver-d8b5aaf433c6b018.so(+0x1299b75)[0x7fb817951b75]
  /checkout/obj/build/x86_64-unknown-linux-gnu/stage1/lib/librustc_driver-d8b5aaf433c6b018.so(_RNvNtCsaXHAFpfKoep_15rustc_interface9interface15parse_check_cfg+0x97)[0x7fb817a4c3e7]
  /checkout/obj/build/x86_64-unknown-linux-gnu/stage1/lib/librustc_driver-d8b5aaf433c6b018.so(_RNvMs_Csd2UaMTeiMW3_17rustc_driver_implNtB4_11RunCompiler3run+0x2c1)[0x7fb8178799f1]
  /checkout/obj/build/x86_64-unknown-linux-gnu/stage1/lib/librustc_driver-d8b5aaf433c6b018.so(+0x11e31fb)[0x7fb81789b1fb]
  /checkout/obj/build/x86_64-unknown-linux-gnu/stage1/lib/librustc_driver-d8b5aaf433c6b018.so(_RNvCsd2UaMTeiMW3_17rustc_driver_impl4main+0x165)[0x7fb817882235]
  /checkout/obj/build/x86_64-unknown-linux-gnu/stage1/bin/rustc(+0x19d57)[0x55b01deaed57]
  /checkout/obj/build/x86_64-unknown-linux-gnu/stage1/bin/rustc(+0x19db3)[0x55b01deaedb3]
  /checkout/obj/build/x86_64-unknown-linux-gnu/stage1/bin/rustc(+0x19da9)[0x55b01deaeda9]
  /checkout/obj/build/x86_64-unknown-linux-gnu/stage1/lib/libstd-5edd1f0c45987863.so(_ZN3std2rt19lang_start_internal17ha6231f3553cf162cE+0x42c)[0x7fb81a7cd34c]
  /checkout/obj/build/x86_64-unknown-linux-gnu/stage1/bin/rustc(+0x19d88)[0x55b01deaed88]
  /lib64/libc.so.6(__libc_start_main+0xf5)[0x7fb815ce4555]
  /checkout/obj/build/x86_64-unknown-linux-gnu/stage1/bin/rustc(+0x19c89)[0x55b01deaec89]
  [RUSTC-TIMING] ___ test:false 0.208
  rustc exited with signal: 11 (SIGSEGV) (core dumped)
stage-build INFO: Section `Stage 1 (LLVM PGO) > Build rustc and LLVM` ended: FAIL (675.92s)
stage-build INFO: Section `Stage 1 (LLVM PGO)` ended: FAIL (675.92s)
stage-build ERROR: The multi-stage build has failed
stage-build INFO: Timer results
---
Total duration:                          11m 16s
------------------------------------------------
root INFO: Free disk space: 508.55 GiB out of total 581.32 GiB (12.51% used)
Traceback (most recent call last):
  File "../src/ci/stage-build.py", line 839, in <module>
    raise e
  File "../src/ci/stage-build.py", line 836, in <module>
    execute_build_pipeline(timer, pipeline, build_args)
  File "../src/ci/stage-build.py", line 760, in execute_build_pipeline
    LLVM_PROFILE_DIR=str(pipeline.llvm_profile_dir_root() / "prof-%p")
  File "../src/ci/stage-build.py", line 571, in build_rustc
    cmd(arguments, env=env)
  File "../src/ci/stage-build.py", line 452, in cmd
    return subprocess.run(args, env=environment, check=True)
  File "/usr/lib64/python3.6/subprocess.py", line 438, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['/usr/bin/python3', '/checkout/x.py', 'build', '--target', 'x86_64-unknown-linux-gnu', '--host', 'x86_64-unknown-linux-gnu', '--stage', '2', 'library/std', '--llvm-profile-generate']' returned non-zero exit status 1.
```

</details>

--- a/compiler/rustc_data_structures/src/sync.rs
+++ b/compiler/rustc_data_structures/src/sync.rs
@@ -107,6 +107,14 @@ cfg_if! {
             }
         }
 
+        impl Atomic<bool> {
+            pub fn fetch_or(&self, val: bool, _: Ordering) -> bool {
+                let result = self.0.get() | val;
+                self.0.set(val);
+                result
+            }
+        }
+
         impl<T: Copy + PartialEq> Atomic<T> {
             #[inline]
             pub fn compare_exchange(&self,

--- a/compiler/rustc_data_structures/src/sync/vec.rs
+++ b/compiler/rustc_data_structures/src/sync/vec.rs
@@ -2,6 +2,7 @@ use std::marker::PhantomData;
 
 use rustc_index::vec::Idx;
 
+#[derive(Default)]
 pub struct AppendOnlyIndexVec<I: Idx, T: Copy> {
     #[cfg(not(parallel_compiler))]
     vec: elsa::vec::FrozenVec<T>,
@@ -40,6 +41,7 @@ impl<I: Idx, T: Copy> AppendOnlyIndexVec<I, T> {
     }
 }
 
+#[derive(Default)]
 pub struct AppendOnlyVec<T: Copy> {
     #[cfg(not(parallel_compiler))]
     vec: elsa::vec::FrozenVec<T>,
@@ -57,11 +59,14 @@ impl<T: Copy> AppendOnlyVec<T> {
         }
     }
 
-    pub fn push(&self, val: T) {
+    pub fn push(&self, val: T) -> usize {
+        #[cfg(not(parallel_compiler))]
+        let i = self.vec.len();
         #[cfg(not(parallel_compiler))]
         self.vec.push(val);
         #[cfg(parallel_compiler)]
-        self.vec.push(val)
+        let i = self.vec.push(val);
+        i
     }
 
     pub fn get(&self, i: usize) -> Option<T> {

--- a/compiler/rustc_metadata/src/rmeta/encoder.rs
+++ b/compiler/rustc_metadata/src/rmeta/encoder.rs
@@ -1712,8 +1712,7 @@ impl<'a, 'tcx> EncodeContext<'a, 'tcx> {
             let stability = tcx.lookup_stability(CRATE_DEF_ID);
             let macros =
                 self.lazy_array(tcx.resolutions(()).proc_macros.iter().map(|p| p.local_def_index));
-            let spans = self.tcx.sess.parse_sess.proc_macro_quoted_spans();
-            for (i, span) in spans.into_iter().enumerate() {
+            for (i, span) in self.tcx.sess.parse_sess.proc_macro_quoted_spans() {
                 let span = self.lazy(span);
                 self.tables.proc_macro_quoted_spans.set_some(i, span);
             }

--- a/compiler/rustc_parse/src/parser/mod.rs
+++ b/compiler/rustc_parse/src/parser/mod.rs
@@ -1541,10 +1541,8 @@ pub(crate) fn make_unclosed_delims_error(
 }
 
 pub fn emit_unclosed_delims(unclosed_delims: &mut Vec<UnmatchedDelim>, sess: &ParseSess) {
-    let _ = sess.reached_eof.compare_exchange(
-        false,
+    let _ = sess.reached_eof.fetch_or(
         unclosed_delims.iter().any(|unmatched_delim| unmatched_delim.found_delim.is_none()),
-        Ordering::Relaxed,
         Ordering::Relaxed,
     );
     for unmatched in unclosed_delims.drain(..) {

--- a/compiler/rustc_session/src/parse.rs
+++ b/compiler/rustc_session/src/parse.rs
@@ -330,10 +330,7 @@ impl ParseSess {
     pub fn proc_macro_quoted_spans(&self) -> impl Iterator<Item = (usize, Span)> + '_ {
         // This is equivalent to `.iter().copied().enumerate()`, but that isn't possible for
         // AppendOnlyVec, so we resort to this scheme.
-        (0..)
-            .map(|i| (i, self.proc_macro_quoted_spans.get(i)))
-            .take_while(|(_, o)| o.is_some())
-            .filter_map(|(i, o)| Some((i, o?)))
+        self.proc_macro_quoted_spans.iter_enumerated()
     }
 
     #[track_caller]

--- a/compiler/rustc_session/src/parse.rs
+++ b/compiler/rustc_session/src/parse.rs
@@ -8,7 +8,7 @@ use crate::lint::{
 };
 use rustc_ast::node_id::NodeId;
 use rustc_data_structures::fx::{FxHashMap, FxHashSet, FxIndexSet};
-use rustc_data_structures::sync::{AtomicBool, Lock, Lrc};
+use rustc_data_structures::sync::{AppendOnlyVec, AtomicBool, Lock, Lrc};
 use rustc_errors::{emitter::SilentEmitter, ColorConfig, Handler};
 use rustc_errors::{
     fallback_fluent_bundle, Diagnostic, DiagnosticBuilder, DiagnosticId, DiagnosticMessage,
@@ -219,7 +219,7 @@ pub struct ParseSess {
     pub assume_incomplete_release: bool,
     /// Spans passed to `proc_macro::quote_span`. Each span has a numerical
     /// identifier represented by its position in the vector.
-    pub proc_macro_quoted_spans: Lock<Vec<Span>>,
+    pub proc_macro_quoted_spans: AppendOnlyVec<Span>,
     /// Used to generate new `AttrId`s. Every `AttrId` is unique.
     pub attr_id_generator: AttrIdGenerator,
 }
@@ -324,13 +324,16 @@ impl ParseSess {
     }
 
     pub fn save_proc_macro_span(&self, span: Span) -> usize {
-        let mut spans = self.proc_macro_quoted_spans.lock();
-        spans.push(span);
-        return spans.len() - 1;
+        self.proc_macro_quoted_spans.push(span)
     }
 
-    pub fn proc_macro_quoted_spans(&self) -> Vec<Span> {
-        self.proc_macro_quoted_spans.lock().clone()
+    pub fn proc_macro_quoted_spans(&self) -> impl Iterator<Item = (usize, Span)> + '_ {
+        // This is equivalent to `.iter().copied().enumerate()`, but that isn't possible for
+        // AppendOnlyVec, so we resort to this scheme.
+        (0..)
+            .map(|i| (i, self.proc_macro_quoted_spans.get(i)))
+            .take_while(|(_, o)| o.is_some())
+            .filter_map(|(i, o)| Some((i, o?)))
     }
 
     #[track_caller]

